### PR TITLE
Knit progress dialog

### PR DESCRIPTION
**Pull request by t0mpr1c3**
Referencing https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/226 and https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/257

A window opens displaying row, color, and needle selection information as knitting proceeds.

It will need a good amount of testing to refine the information displayed and check that the UI is robust. I think there is something wrong with how it re-initializes because the dialog box works pretty well first time around, at least in the simulation mode, but there are frequent glitches when restarting the knitting process.

### CODE REVIEW

**Comment by chris007de** [comment]
Thanks a lot for another useful addition to the project!
I love the representation of the colors, that's really a huge improvement.

Some thoughts on it:
- Why a separate Window?
  - When clicking "Configure", it steals the attention from the main window, but we still have to click "Knit" to actually start knitting. That's not very intuitive.
  - The window shows also when I configure a non-allowed setting (singlebed with 3 colors)
  - Wouldn't it also be OK to split the main window horizontally and show the content of the progress window there? (e.g. Top: Overview, Bottom: Knitting Progress)

- Automatic scrolling is a must-have, I guess (at least on my system, it does not scroll automatically on addition of a new pass)
- When I cancelled knitting and start another knitting, the program is stuck . It seems that the knitting state (reqLine etc.) are not reset properly on cancel. (When cancelling at row 2, the first requested row at the next knitting is row 3). Not sure if related to the progress window, though.
- Also not sure if this is related to the progress window: In multicolor, if we do not use a certain color in a row, couldn't we skip the passes for this color?
- What's the meaning of the small triangles inside the color blocks?

File: src/main/python/ayab/ayab.py [hunk 1]
```diff
@@ -48,6 +52,12 @@
 
 MACHINE_WIDTH = 200
 
+# TODO move to generic configuration
+MACHINE_WIDTH = 200
```
Comment (chris007de) [inline src/main/python/ayab/ayab.py:56]: Duplicate of line 53

File: src/main/python/ayab/ayab.py [hunk 2]
```diff
@@ -65,18 +75,51 @@
...
+
+def rgb2array(color):
+    return np.array([color // 0x1000, (color // 0x100) & 0xFF, color & 0xFF])
+
+def greyscale(rgb):
+    return np.dot([.299, .587, .114], rgb)
+
+def contrast_color(color):
+    return [0x000000,0xFFFFFF][greyscale(rgb2array(color)) < 0x80]
+
```
Comment (chris007de) [inline src/main/python/ayab/ayab.py:110]: Probably worth to put these functions as methods into the ayab_progress class.

File: src/main/python/ayab/plugins/ayab_plugin/ayab_communication_mockup.py [hunk 3]
```diff
@@ -54,6 +58,20 @@ def update(self) -> list:
             self.__line_count += 1
             self.__line_count %= 256
             self.__rxMsgList.append(reqLine)
+            if self.__delay:
+                sleep(1) # wait for knitting progress dialog to update
+
+            # nasty kludge: set condition to True for line-by-line debugging
+            if False:
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_communication_mockup.py:65]: Maybe decide this with a parameter in the constructor, so we can choose the behavior with a parameter in the GUI in the future.

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 4]
```diff
@@ -472,46 +537,46 @@ def knit(self, pImage, pOptions):
...
-        elif self._current_state == KnittingState.INIT:
+        elif self.__current_state == KnittingState.INIT:
             if rcvMsg == 'cnfInfo':
-                if rcvParam == self._API_VERSION:
-                    self._current_state = KnittingState.WAIT_FOR_INIT
+                if rcvParam == self.__API_VERSION:
+                    self.__current_state = KnittingState.WAIT_FOR_INIT
                     result = AYABControlKnitResult.WAIT_FOR_INIT
                 else:
                     self.__logger.error("wrong API version: " +
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:548]: Unrelated: Sooner or later, we should show the Firmware Version here, as the user is not aware of an API version at all.

File: src/main/python/ayab/ayab.py [hunk 5]
```diff
@@ -160,6 +213,41 @@ def updateStatus(self, hall_l, hall_r, carriage_type, carriage_position):
...
+                # TODO: hints, notes, memos
+                # self.kp.area.ensureWidgetVisible(w0)
+            # graph line of stitches
+            for c in range(len(progress.bits)):
+                wc = stitch(progress.color, 0, progress.bits[c], progress.alt_color)
+                self.kp.grid.addWidget(wc, self.kp.row, 6 + c)
+            # if not infinite repeats, update progress bar
+            if progress.repeats < 0 and progress.total_rows > 0:  
+                self.kp.pd.setValue(100 * (row + 1) / progress.total_rows)
+            self.kp.pd.update()
```
Comment (chris007de) [inline src/main/python/ayab/ayab.py:249]: Move logic into ayab_progress method?

Comment (t0mpr1c3) [reply src/main/python/ayab/ayab.py:249]: Yes

File: src/main/python/ayab/ayab.py (hunk 1 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/ayab.py:56]: Thanks. Eventually I'm going to move all the constants into a single module as you suggested.

File: src/main/python/ayab/ayab.py (hunk 2 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/ayab.py:110]: Yes

File: src/main/python/ayab/plugins/ayab_plugin/ayab_communication_mockup.py (hunk 3 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_communication_mockup.py:65]: Yes

**Comment by t0mpr1c3** [comment]
Thanks Chris -- these are very helpful comments

1. It probably should not be in a separate window, but I don't have a very good sense of how to organize the viewing area at the moment. The horizontal split is a sensible suggestion. Probably it should go at the bottom near where the progress bar is at the moment. And it should probably emerge when you press "Knit" instead of "configure".

2. I struggled with implementing automatic scrolling. I will figure it out eventually.

3. Yes, this is a particularly bad bug. Something does not reset properly on cancel.

4. Yes, as far as I can tell this is missing from the knitting logic entirely. The only suggestion that I have seen is not to design patterns with bank rows! (I think that's in one of the DBJ manual pages.) I suggest that we raise a new issue to deal with that.

5. It's an experimental feature, we can get rid of them. The little dots are intended to represent knit stitches (as opposed to purl stitches which are usually charted with a vertical line, yarn overs with a circle, etc). Since we do not currently implement any knowledge of different stitch types there is no reason to show these symbols at the moment. (I would like to change that in future so that it is possible to knit in both different colours and different textures. At the moment AYAB barely supports lace knitting. But it will be a big change, and I am sure it won't happen any time soon.)

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 6]
```diff
@@ -181,88 +215,94 @@ def __cnfLine(self, lineNumber):
...
 
-            # TODO implement CRC8
-            crc8 = 0x00
-
             # send line to machine
-            if self._infRepeat:
-                self.__ayabCom.cnf_line(requestedLine, bits.tobytes(), 0, crc8)
-            else:
-                self.__ayabCom.cnf_line(requestedLine, bits.tobytes(), lastLine, crc8)
+            self.__ayabCom.cnf_line(requestedLine, bits.tobytes(), lastLine & self.__infRepeat)
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:225]: AyabCommunicationMockup.cnf_line() still expects (lineNumber, lineData, flags, crc8). Please remove the crc8 parameter so we have the same interface for both variants

Comment (t0mpr1c3) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:210]: OK

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 7]
```diff
@@ -54,121 +60,136 @@ class AYABControlKnitResult(Enum):
...
-        elif self._knitting_mode == KnittingMode.MIDDLECOLORSTWICE_RIBBER.value and self._numColors >= 2:
-            self._knit_func = self._middlecoltwice
-        elif self._knitting_mode == KnittingMode.HEARTOFPLUTO_RIBBER.value and self._numColors >= 2:
-            self._knit_func = self._heartofpluto
-        elif self._knitting_mode == KnittingMode.CIRCULAR_RIBBER.value: # not restricted to 2 colors
-            self._knit_func = self._circular_ribber
-        else:
-            self.__logger.error("Fallthrough error in __get_knit_func: invalid knitting options")
-            return False # knit function not found
+        if not self.__knitting_mode.good_ncolors(self.__numColors):
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:164]: self.__knitting_mode needs a cast to KnittingMode().

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 8]
```diff
@@ -181,88 +202,83 @@ def __cnfLine(self, lineNumber):
...
             lineNumber += BLOCK_LENGTH * self.__lineBlock
 
-            imgHeight = self._image.imgHeight()
-        
-            # work out which line of data to send
-            imgRow, color, indexToSend, sendBlankLine, lastLine = self._knit_func(lineNumber, imgHeight)
-            
-            # create bitarray
+            # get data for next line of knitting
+            imgRow, color, indexToSend, sendBlankLine, lastLine = self.__knit_func(lineNumber)
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:206]: Trying to knit singlebed yields: `TypeError: _singlebed() missing 1 required positional argument: 'lineNumber'`.
Possible to add a unit test for cnf_line?

**Comment by t0mpr1c3** [comment]
Thanks, Chris! I will get around to the making unit tests for the new methods at some point. Right now I am trying to fix up the UI. I switched from a pop-up dialog box to a split window as you suggested. It needs a few tweaks but overall I think it is much better. Also I think I have fixed the problem I was having with the auto scrolling. It still crashes when you try to knit again after cancelling, I am trying to track that down next.

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py (hunk 7 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_control.py]: Done

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py (hunk 8 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_control.py]: Don't we already have tests for `cnf_line()` in `test_communication.py` and `test_communication_mockup.py`?

**Comment by t0mpr1c3** [comment]
Chris -- is restarting knitting where you left off a bug, or is it supposed to be a feature? If it is a feature, I suggest that it should be implemented by setting the "start line" form entry to the last line that was actually completed plus one.

**Comment by chris007de** [comment]
> Chris -- after cancelling knitting, is restarting where you left off a bug, or is it supposed to be a feature?

As far as I remember, it should start from scratch and not preserve the line - so I would call it a bug (because we do not have a UI element to actually reset it to start from scratch, so it should happen automatically)

**Comment by t0mpr1c3** [comment]
I started a new issue to address the program flow bugs: https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/338

**Comment by t0mpr1c3** [comment]
I think I fixed the bug. The problem was that after configure, knit, cancel, configure, knit, the FSM was never reset to SETUP and was still in OPERATE.

**Comment by t0mpr1c3** [comment]
Chris, this is ready for further review when you get a chance.

**Comment by chris007de** [comment]
Great work...only one drawback: The unit tests (especially test_req_line) currently take about 4:30 minutes to complete. We should definitely address that :)
I'll merge this for now, polishing/issues can be addressed separately.
Thanks a lot!

**Comment by t0mpr1c3** [comment]
Thanks. I have no idea why the tests are so slow.

**Pull request merged**